client/modules/sheaf-core/src/extensions/indent-hack.ts

```diff
@@ -0,0 +1,64 @@
+/**
+ * @file Exports the "indent-hack" for CodeMirror.
+ * This enables elegant, tab-preserving line-wrapping in the editor.
+ */
+
+import { EditorState } from "@codemirror/state"
+import {
+  EditorView,
+  ViewPlugin,
+  ViewUpdate,
+  Decoration,
+  DecorationSet
+} from "@codemirror/view"
+import { RangeSetBuilder } from "@codemirror/rangeset"
+import type { Line } from "@codemirror/text"
+
+const WHITESPACE_REGEX = /^\s+/
+
+function indentDeco(view: EditorView) {
+  // get every line of the visible ranges
+  const lines = new Set<Line>()
```

Comment:
We don't want repeats. So, if we're checking line 15, and we somehow get line 15 again, we don't want to indent line 15 twice because it was in the list twice.